Objects/frameobject.c

```diff
@@ -42,6 +42,150 @@ frame_getlineno(PyFrameObject *f, void *closure)
+            if (setup_op == SETUP_FINALLY) {
+                /* This is the start of a 'finally' block.
+                 * It will end with RERAISE (if a 'try..finally' block)
+                 * or WITH_CLEANUP_FINISH (if a 'with' block).
+                 */
+                in_finally[blockstack_top-1] = 1;
+            }
+            else {
+                blockstack_top--;
+            }
+            break;
+
+        case RERAISE:
+            if (blockstack_top > 0) {
+                setup_op = code[blockstack[blockstack_top-1]];
+                if (setup_op == SETUP_FINALLY) {
+                    /* This is the end of a 'finally' block. */
+                    blockstack_top--;
+                }
+                /* RERAISE is also used by SETUP_EXCEPT, but we don't care. */
+            }
+            break;
+        case WITH_CLEANUP_FINISH:
+            /* This is the end of a 'finally' block */
+            assert(blockstack_top > 0);
+            assert(code[blockstack[blockstack_top-1]] == SETUP_FINALLY);
+            blockstack_top--;
+            break;
+        }
+
+        /* For the addresses we're interested in, see whether they're
+         * within a 'finally' block and if so, remember the address
+         * of the SETUP_FINALLY. */
+        if (addr == first_addr || addr == second_addr) {
+            int i = 0;
+            int setup_addr = -1;
+            for (i = blockstack_top-1; i >= 0; i--) {
+                if (in_finally[i]) {
+                    setup_addr = blockstack[i];
+                    break;
+                }
+            }
+
+            if (setup_addr != -1) {
+                if (addr == first_addr) {
+                    *first_setup_addr = setup_addr;
+                }
+
+                if (addr == second_addr) {
+                    *second_setup_addr = setup_addr;
+                }
+            }
+        }
+    }
+
+    /* Verify that the blockstack tracking code didn't get lost. */
+    assert(blockstack_top == 0);
+}
+
+
+/*
+ * Compute and return the 'for' loop nesting levels
+ * for *first_addr* and *second_addr*.
+ */
+static void
+compute_for_loop_levels(unsigned char *code, Py_ssize_t code_len,
+                        int first_addr, int second_addr,
+                        int *first_loop_level, int *second_loop_level)
+{
+    int forstack[CO_MAXBLOCKS];
+    int forstack_top = 0;
+    int addr = 0;
+
+    forstack_top = 0;
+    *first_loop_level = -1;
+    *second_loop_level = -1;
+
+    for (addr = 0; addr < code_len; addr += sizeof(_Py_CODEUNIT)) {
+        unsigned char op = code[addr];
+        unsigned char oparg;
+        int target_addr;
+
+        switch (op) {
+        case FOR_ITER:
+            /* Store the loop end in forstack[] */
+            oparg = code[addr + 1];
+            target_addr = addr + oparg + 2;
+            assert(target_addr < code_len);
+            forstack[forstack_top++] = target_addr;
```

Comment:
> The jump is forbidden if (addr < first_addr && first_addr < target_addr) != (addr < second_addr && second_addr < target_addr) for any 'for' block.

I don't think this is true. It is actually allowed to jump *out* of a `for` loop, just not *into*.